### PR TITLE
Added LocationManager to DEMO, now app moves focus to the current gps…

### PR DESCRIPTION
**Pull request by turcato-niccolo**
… position when app opens.

Added Support for FusedLocationClient API (Android Developer suggested)

### CODE REVIEW

File: demo/src/main/java/com/eis/geoCalendar/demo/MapsActivity.java [hunk 1]
```diff
@@ -19,6 +21,7 @@
 import com.google.android.gms.maps.model.LatLng;
 import com.google.android.gms.maps.model.Marker;
 import com.google.android.gms.maps.model.MarkerOptions;
+import com.eis.geoCalendar.demo.Localization.*;
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/MapsActivity.java]: Codacy Issue found: [Using the '.*' form of import should be avoided - com.eis.geoCalendar.demo.Localization.*.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 12
           

Complexity increasing per file
==============================
- demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java  1
- demo/src/main/java/com/eis/geoCalendar/demo/MoveMapCommand.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932&bid=16036118)

File: demo/src/main/java/com/eis/geoCalendar/demo/MoveMapCommand.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import com.google.android.gms.maps.GoogleMap;
+import com.google.android.gms.maps.model.LatLng;
+
+/**
+ * Used to move the Map to the defined Location, with the max possible zoom
+ */
+public class MoveMapCommand implements Command<Location> {
+    private GoogleMap googleMap;
+
+    public MoveMapCommand(GoogleMap map){
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/MoveMapCommand.java]: Codacy Issue found: ['{' is not preceded with whitespace.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    //NOTE: concat latitude,longitude
+
+    private Context currentContext;
+    private LocationRequest locationRequest;
+    private PendingIntent locationIntent;
+    private Location mLastLocation;
+
+    private FusedLocationProviderClient mFusedLocationClient;
+
+    public LocationManager(Context applicationContext){
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: ['{' is not preceded with whitespace.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+import static com.google.android.gms.location.LocationRequest.PRIORITY_HIGH_ACCURACY;
+
+/**
+ * @author Turcato
+ *
+ * TODO: this is here temporarely just for the demo
+ */
+public class LocationManager {
+    private static final String[] permissions = {
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: [Name 'permissions' must match pattern '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                            mLastLocation is used directly here because once out of OnComplete
+                            the Location isn't available and the variable that contains it
+                            becomes null
+                            */
+                        } else if (!task.isSuccessful()) {
+                            Log.d(LocationManagerTag, "Task<Location> not successful");
+                        } else if (task.getResult() == null) {
+                            Log.d(LocationManagerTag, "Task<Location> result is null");
+                        }
+                        Log.d(LocationManagerTag, "End of OnComplete " +mLastLocation.toString());
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: ['+' is not followed by whitespace.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
+package com.eis.geoCalendar.demo.Localization;
+
+import android.Manifest;
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.location.Location;
+import android.net.Uri;
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: [Unused import - android.net.Uri.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+                mFusedLocationClient.getLastLocation().addOnCompleteListener(new OnCompleteListener<Location>() {
+                    @Override
+                    public void onComplete(@NonNull Task<Location> task) {
+                        Log.d(LocationManagerTag, "Completed lastLocation");
+                        Log.d(LocationManagerTag, "Task<Location> successful " +  task.isSuccessful());
+
+                        if (task.isSuccessful() && task.getResult() != null) {
+                            mLastLocation = task.getResult();
+                            Log.d(LocationManagerTag, "Victory!" +mLastLocation.toString());
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: ['+' is not followed by whitespace.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+
+import static com.google.android.gms.location.LocationRequest.PRIORITY_HIGH_ACCURACY;
+
+/**
+ * @author Turcato
+ *
+ * TODO: this is here temporarely just for the demo
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: [Comment matches to-do format 'TODO:'.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
+package com.eis.geoCalendar.demo.Localization;
+
+import android.Manifest;
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: [Unused import - android.content.Intent.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+public class LocationManager {
+    private static final String[] permissions = {
+            Manifest.permission.ACCESS_COARSE_LOCATION,
+            Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.READ_PHONE_STATE,
+    };
+
+
+    private final String LocationManagerTag = "LocationManagerTag";
+    private final String MAPS_START_URL = "https://www.google.com/maps/search/?api=1&query=";
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: [This final field could be made static](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+ */
+public class LocationManager {
+    private static final String[] permissions = {
+            Manifest.permission.ACCESS_COARSE_LOCATION,
+            Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.READ_PHONE_STATE,
+    };
+
+
+    private final String LocationManagerTag = "LocationManagerTag";
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: [This final field could be made static](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java (hunk 10 above)
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: [Avoid unused private fields such as 'MAPS_START_URL'.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- demo/src/main/java/com/eis/geoCalendar/demo/MoveMapCommand.java  1
- demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932&bid=16036118)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+ */
+public class LocationManager {
+    private static final String[] PERMISSIONS = {
+            Manifest.permission.ACCESS_COARSE_LOCATION,
+            Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.READ_PHONE_STATE,
+    };
+
+
+    private final static String LocationManagerTag = "LocationManagerTag";
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: [Name 'LocationManagerTag' must match pattern '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+
+import static com.google.android.gms.location.LocationRequest.PRIORITY_HIGH_ACCURACY;
+
+/**
+ * @author Turcato
+ * <p>
+ * TODO: this is here temporarely just for the demo
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: [Comment matches to-do format 'TODO:'.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java (hunk 12 above)
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java]: Codacy Issue found: ['static' modifier out of order with the JLS suggestions.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- demo/src/main/java/com/eis/geoCalendar/demo/MoveMapCommand.java  1
- demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932&bid=16036118)

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java (hunk 13 above)
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java:26]: Codacy Issue found: [Comment matches to-do format 'TODO:'.](https://app.codacy.com/manual/CremaLuca/geoCalendar/pullRequest?prid=4824932)

**Review by CremaLuca — COMMENTED** [review]
1) This is what @tonin-alessandra is working on too, but she's making it continuos
2) What in the world is a `Command`? Why didn't you use a common listener?

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+import com.google.android.gms.tasks.OnCanceledListener;
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+
+import static com.google.android.gms.location.LocationRequest.PRIORITY_HIGH_ACCURACY;
+
+/**
+ * @author Turcato
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java:24]: Tags are usually placed at the end of the specifications

File: demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        currentContext = applicationContext;
+    }
+
+    /***
+     * Method that gets the last Location available of the device, and executes the imposed command
+     * calling command.execute(foundLocation)
+     *
+     * @param command object of a class that implements interface Command
+     */
+    public void getLastLocation(final Command<Location> command) {
```
Comment (CremaLuca) [inline demo/src/main/java/com/eis/geoCalendar/demo/Localization/LocationManager.java:55]: Why are you not using a `GetLastLocationListener` or something like that? This `Command` looks like a delegate.

**Comment by turcato-niccolo** [comment]
> 1. This is what @tonin-alessandra is working on too, but she's making it continuos

Good, it can be replaced as soon as it's available 

> 2. What in the world is a `Command`? Why didn't you use a common listener?

It's a design pattern, and the reason it's here is because of bad design, I had in mind to do a complete refactoring of this LocationManager, but if someone else is already doing the same thing, even better.
I needed it just for the demo.

**Review by CremaLuca — COMMENTED** [review]
Some comments I did a couple hours ago and I forgot to submit.

**Pull request merged**